core/src/main/java/jenkins/install/SetupWizard.java

```diff
@@ -245,4 +270,170 @@ public String getRequestURI() {
+                    categoryIterator.remove();
+                }
+            }
+        }
+        return pluginCategories;
+    }
+    
+    /**
+     * Provides the list of platform plugin updates from the last time
+     * the upgrade was run
+     */
+    /*package*/ JSONArray getPlatformPluginUpdates() {
+        try {
+            return getPlatformPluginsForUpdate(getCurrentLevel(), Jenkins.getVersion());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+    
+    /**
+     * Returns the initial plugin list in JSON format
+     */
+    @Restricted(DoNotUse.class) // WebOnly
+    public HttpResponse doPlatformPluginList() throws IOException {
+        jenkins.install.SetupWizard setupWizard = Jenkins.getInstance().getSetupWizard();
+        if (setupWizard != null) {
+            if (InstallState.UPGRADE.equals(Jenkins.getInstance().getInstallState())) {
+                JSONArray initialPluginData = getPlatformPluginUpdates();
+                if(initialPluginData != null) {
+                    return hudson.util.HttpResponses.okJSON(initialPluginData);
+                }
+            } else {
+                JSONArray initialPluginData = getPlatformPluginList();
+                if(initialPluginData != null) {
+                    return hudson.util.HttpResponses.okJSON(initialPluginData);
+                }
+            }
+        }
+        return hudson.util.HttpResponses.okJSON();
+    }
+    
+    /*package*/ void completeUpgrade(Jenkins jenkins) throws IOException {
+        setCurrentLevel(Jenkins.getVersion());
+        jenkins.getInstallState().proceedToNextState();
+    }
+    
+    /*package*/ void setCurrentLevel(VersionNumber v) throws IOException {
+        FileUtils.writeStringToFile(getUpdateStateFile(), v.toString());
+    }
+    
+    /**
+     * File that captures the state of upgrade.
+     *
+     * This file records the vesrion number that the installation has upgraded to.
+     */
+    /*package*/ File getUpdateStateFile() {
+        return new File(Jenkins.getInstance().getRootDir(),"jenkins.install.UpgradeWizard.state");
+    }
+
+    /**
+     * What is the version the upgrade wizard has run the last time and upgraded to?
+     */
+    public VersionNumber getCurrentLevel() throws IOException {
+        VersionNumber from = new VersionNumber("1.0");
+        File state = getUpdateStateFile();
+        if (state.exists()) {
+            from = new VersionNumber(defaultIfBlank(readFileToString(state), "1.0").trim());
+        }
+        return from;
+    }
+    
+    /**
+     * Gets all the install states
+     */
+    public List<InstallState> getInstallStates() {
+        return InstallState.all();
+    }
+    
+    public String getStartPanel() {
+        return Jenkins.getInstance().getInstallState().getStartPanel(new Provider<String>() {
+            @Override
+            public String get() {
+                return "welcomePanel";
```

Comment:
It could help the extensibility cases if we can change this value via system property.